src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeTests.cs

```diff
@@ -2538,6 +2538,161 @@ class C
+        public void TestRemoveWithoutEOL_KeepEndOfLine()
+        {
+            var cu = SyntaxFactory.ParseCompilationUnit(@"class A { } class B { } // test");
+
+            var m = cu.DescendantNodes().OfType<TypeDeclarationSyntax>().LastOrDefault();
+            Assert.NotNull(m);
+
+            var cu2 = cu.RemoveNode(m, SyntaxRemoveOptions.KeepEndOfLine);
+
+            var text = cu2.ToFullString();
+
+            Assert.Equal("class A { } ", text);
+        }
+
+        [Fact]
+        [WorkItem(22924, "https://github.com/dotnet/roslyn/issues/22924")]
+        public void TestRemoveBadDirectiveWithoutEOL_KeepEndOfLine_KeepDirectives()
+        {
+            var cu = SyntaxFactory.ParseCompilationUnit(@"class A { } class B { } #endregion");
+
+            var m = cu.DescendantNodes().OfType<TypeDeclarationSyntax>().LastOrDefault();
+            Assert.NotNull(m);
+
+            var cu2 = cu.RemoveNode(m, SyntaxRemoveOptions.KeepEndOfLine | SyntaxRemoveOptions.KeepDirectives);
+
+            var text = cu2.ToFullString();
+
+            Assert.Equal("class A { } #endregion", text);
+        }
+
+        [Fact]
+        [WorkItem(22924, "https://github.com/dotnet/roslyn/issues/22924")]
+        public void TestRemoveDocument_KeepEndOfLine()
+        {
+            var cu = SyntaxFactory.ParseCompilationUnit(@"
+#region A
+class A 
+{ } 
+#endregion");
+
+            var cu2 = cu.RemoveNode(cu, SyntaxRemoveOptions.KeepEndOfLine);
+
+            Assert.Null(cu2);
+        }
+
+        [Fact]
+        [WorkItem(22924, "https://github.com/dotnet/roslyn/issues/22924")]
+        public void TestRemoveFirstParameter_KeepEndOfLine()
+        {
+            var cu = SyntaxFactory.ParseCompilationUnit(@"
+class C
+{
+void M(
+// before a
+int a,
+// after a
+// before b
+int b
+// after b)
+{
+}
+}");
+
+            var expectedText = @"
+class C
+{
+void M(
+
+// after a
+// before b
+int b
+// after b)
+{
+}
+}";
+
+            var m = cu.DescendantNodes().OfType<ParameterSyntax>().FirstOrDefault();
+            Assert.NotNull(m);
+
+            var cu2 = cu.RemoveNode(m, SyntaxRemoveOptions.KeepEndOfLine);
+
+            var text = cu2.ToFullString();
+
+            Assert.Equal(expectedText, text);
+        }
+
+        [Fact]
+        [WorkItem(22924, "https://github.com/dotnet/roslyn/issues/22924")]
+        public void TestRemoveLastParameter_KeepEndOfLine()
+        {
+            var cu = SyntaxFactory.ParseCompilationUnit(@"
+class C
+{
+void M(
+// before a
+int a,
+// after a
+// before b
+int b
+// after b)
+{
+}
+}");
+            var expectedText = @"
+class C
+{
+void M(
+// before a
+int a
+// after b)
+{
+}
+}";
+
+            var m = cu.DescendantNodes().OfType<ParameterSyntax>().LastOrDefault();
+            Assert.NotNull(m);
+
+            var cu2 = cu.RemoveNode(m, SyntaxRemoveOptions.KeepEndOfLine);
+
+            var text = cu2.ToFullString();
+
+            Assert.Equal(expectedText, text);
+        }
```

Comment:
>} [](start = 8, length = 1)

I experimented with reverting most of the changes in SyntaxNodeRemover (I left the new null check and making `HasEndOfLine` smarter). 
In that set up, all the added tests pass but one (`TestRemoveBadDirectiveWithoutEOL_KeepEndOfLine_KeepDirectives`)

Is this because I'm running the tests on windows (so my end-of-lines are already CRLF)?

The purpose of this PR is that previously we would insert a CarriageReturnLineFeed, but now we insert an EOL that matches existing trivia. Is it possible to add tests that illustrate this change (preferably even when running on Windows)?

I'm not sure how to do this, but one possible idea: parse the source as normal, but replace all EOL trivia with alternative sequences (like LineFeed, which I believe is Unix), then invoke API (`RemoveNode`) and verify that no CarriageReturnLineFeed was inserted. Such tests should fail prior to fix.

Reply:
> Is this because I'm running the tests on windows (so my end-of-lines are already CRLF)?

Yes, on Windows all EOLs are CRLF that is why almost all tests passed.

> Is it possible to add tests that illustrate this change (preferably even when running on Windows)?

In the first commit of this PR all tests were tested with CRLF and LF on Windows and Unix by using a "normalizer" function as a parameter ([here](https://github.com/dotnet/roslyn/pull/23165/commits/6d2a399ee58cac07472e554f84ed2e466a199077#diff-0e0e5cee589dd5835bc357a2c0406c46R2447) is an example), but was later removed due to [the comment](https://github.com/dotnet/roslyn/pull/23165#discussion_r189049439) from @CyrusNajmabadi.

Reply:
@CyrusNajmabadi I think you commented on another topic (what to do in absence of EOL clue?). 
My question here was how do we tighten the tests, so that it's clear that no CRLF is injected (whereas we used to).

Reply:
So, my idea there would be to run the test and just check that there's no CRLF in it.  If we think one could be introduced because of the remover adding something in automatically, then we resolve that by setting an option saying "if you need to put something and cannot determine what the right EOL should be, then use \n".  

So, the test would run the removal, stating that the default EOF should be "\n".  it would then run and validate only \n's.  Similarly, the test shoudl run saying the default EOF should be \r\n, and hten validate that in this case, it did get a \r\n because the remover couldn't figure out a good default.

Reply:
(i may be getting confused, or causing more confusion than necessary.  if you want, we can chat in gitter if you think we're talking past each other :)).

Reply:
Ok. i thnk what julien said makes sense.  Let's just go with '1' :)

Reply:
> add an explicit validation to the tests: if the input doesn't contain CRLF, then assert that the output does not contain CRLF

I am not sure how would this validation solve the problem of testing a Unix EOL (LF) on Windows. The input or the output text will not contain a CRLF only when the code is tested on a Unix OS.
Also, this validation is already done by `Assert.Equal` as it checks if both texts have exactly the same characters on the same positions, which would detect if the input has a CRLF and output a LF character. Here is an example:
```
Assert.Equal() Failure
                        ↓ (pos 11)
Expected: \nclass C\n{\n\n}
Actual:   \nclass C\n{\n\r\n}
                        ↑ (pos 11)
```
Probably I am missing something here?

In the meantime I've added a static method that will test both Windows and Unix EOL and used on tests where EOL matters. Now by reverting to the old `SyntaxNodeRemover` you will clearly see that all those tests will fail. If the used static method is not appropriate due to code readability or some other reasons I will revert it and I am open to other solutions.